modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java

```diff
@@ -2464,4 +2610,72 @@ boolean handleResult(long reqId, JdbcResult res) {
             return handled;
         }
     }
+
+    /** Transaction context. */
+    private class TxContext {
+        /** IO to transaction coordinator. */
+        final JdbcThinTcpIo txIo;
+
+        /** Transaction id. */
+        final int txId;
+
+        /** Closed flag. */
+        boolean closed;
+
+        /** Tracked statements to close results on transaction end. */
+        private final Set<JdbcThinStatement> stmts = Collections.newSetFromMap(new IdentityHashMap<>());
+
+        /** */
+        public TxContext(JdbcThinTcpIo txIo, int txId) {
+            assert !autoCommit;
+
+            this.txIo = txIo;
+            this.txId = txId;
+        }
+
+        /** */
+        public void end(boolean commit) throws SQLException {
+            if (closed)
+                return;
+
+            closed = true;
+
+            for (JdbcThinStatement stmt : stmts)
+                stmt.closeResults();
```

Comment:
why not just close?

Reply:
Because, JDBC spec allows to keep statements over commit.